packages/bff/src/services/purposeService.ts

```diff
@@ -1,19 +1,260 @@
-import { PagoPAInteropBeClients } from "../providers/clientProvider.js";
-import { purposeNotFound } from "../model/domain/errors.js";
-import { BffAppContext } from "../utilities/context.js";
+import { bffApi, purposeApi, tenantApi } from "pagopa-interop-api-clients";
+import {
+  AgreementProcessClient,
+  AuthorizationProcessClient,
+  CatalogProcessClient,
+  PurposeProcessClient,
+  TenantProcessClient,
+} from "../providers/clientProvider.js";
+import {
+  agreementNotFound,
+  eServiceDescriptorNotFound,
+  eServiceNotFound,
+  purposeDraftVersionNotFound,
+  purposeNotFound,
+  tenantNotFound,
+} from "../model/domain/errors.js";
+import { BffAppContext, Headers } from "../utilities/context.js";
+import { toBffApiCompactClient } from "../model/domain/apiConverter.js";
+import { isUpgradable } from "../model/modelMappingUtils.js";
+import { getLatestAgreement } from "./agreementService.js";
+import { getAllClients } from "./clientService.js";
+
+export const getCurrentVersion = (
+  purposeVersions: purposeApi.PurposeVersion[]
+): purposeApi.PurposeVersion | undefined => {
+  const statesToExclude: purposeApi.PurposeVersionState[] = [
+    purposeApi.PurposeVersionState.Values.WAITING_FOR_APPROVAL,
+    purposeApi.PurposeVersionState.Values.REJECTED,
+  ];
+  return purposeVersions
+    .filter((v) => !statesToExclude.includes(v.state))
+    .sort(
+      (a, b) =>
+        new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime()
+    )
+    .at(-1);
+};
+
+// eslint-disable-next-line max-params
+async function getPurposes(
+  {
+    purposeClient,
+    eserviceClient,
+    tenantClient,
+    agreementClient,
+    authorizationClient,
+  }: {
+    purposeClient: PurposeProcessClient;
+    eserviceClient: CatalogProcessClient;
+    tenantClient: TenantProcessClient;
+    agreementClient: AgreementProcessClient;
+    authorizationClient: AuthorizationProcessClient;
+  },
+  requesterId: string,
+  filters: {
+    name?: string | undefined;
+    eservicesIds?: string[];
+    consumersIds?: string[];
+    producersIds?: string[];
+    states?: purposeApi.PurposeVersionState[];
+    excludeDraft?: boolean | undefined;
+  },
+  offset: number,
+  limit: number,
+  headers: Headers
+): Promise<bffApi.Purposes> {
+  const purposes = await purposeClient.getPurposes({
+    queries: {
+      ...filters,
+      limit,
+      offset,
+    },
+    headers,
+  });
+
+  const eservices = await Promise.all(
+    toSetToArray(purposes.results.map((p) => p.eserviceId)).map((id) =>
+      eserviceClient.getEServiceById({
+        params: {
+          eServiceId: id,
+        },
+        headers,
+      })
+    )
+  );
+
+  const getTenant = async (id: string): Promise<tenantApi.Tenant> =>
+    tenantClient.tenant.getTenant({
+      params: {
+        id,
+      },
+      headers,
+    });
+  const consumers = await Promise.all(
+    [...new Set(purposes.results.map((p) => p.consumerId))].map(getTenant)
```

Comment:
why are you using `new Set(...)` ?  this array `purposes.results.map((p) => p.consumerId)`  it's not enough?

Reply:
The old scale codebase uses a `.distinct` so here I'm using this trick